src/Microsoft.TemplateEngine.Core/Util/ProcessorState.cs

```diff
@@ -54,31 +74,40 @@ public ProcessorState(Stream source, Stream target, int bufferSize, int flushThr
             CurrentBufferPosition = bom.Length;
             target.Write(bom, 0, bom.Length);
 
-            Dictionary<Encoding, OperationTrie> byEncoding;
-            if(!TrieLookup.TryGetValue(operationProviders, out byEncoding))
+            Dictionary<Encoding, Trie<OperationTerminal>> byEncoding;
+            if (!TrieLookup.TryGetValue(operationProviders, out byEncoding))
             {
-                TrieLookup[operationProviders] = byEncoding = new Dictionary<Encoding, OperationTrie>();
+                TrieLookup[operationProviders] = byEncoding = new Dictionary<Encoding, Trie<OperationTerminal>>();
             }
 
-            if (!byEncoding.TryGetValue(encoding, out _trie))
+            if (!byEncoding.TryGetValue(encoding, out Trie<OperationTerminal> trie))
             {
-                List<IOperation> operations = new List<IOperation>(operationProviders.Count);
+                trie = new Trie<OperationTerminal>();
 
                 for (int i = 0; i < operationProviders.Count; ++i)
                 {
                     IOperation op = operationProviders[i].GetOperation(encoding, this);
+
                     if (op != null)
                     {
-                        operations.Add(op);
+                        for (int j = 0; j < op.Tokens.Count; ++j)
+                        {
+                            if (op.Tokens[j] != null)
+                            {
+                                trie.AddPath(op.Tokens[j].Value, new OperationTerminal(op, j, op.Tokens[j].Length, op.Tokens[j].Start, op.Tokens[j].End));
+                            }
+                        }
                     }
                 }
 
-                byEncoding[encoding] = _trie = OperationTrie.Create(operations);
+                byEncoding[encoding] = trie;
             }
 
-            if (bufferSize < _trie.MaxLength && !sizedToStream)
+            _trie = new TrieEvaluator<OperationTerminal>(trie);
+
+            if (bufferSize < _trie.MaxLength + 1)
             {
-                byte[] tmp = new byte[_trie.MaxLength];
+                byte[] tmp = new byte[_trie.MaxLength + 1];
```

Comment:
The sizedToStrream check was removed with this PR. Wondering if there could be problems if the entire source stream is shorter than the _trie.MaxLength + 1.